charging at start of iteration in Residential stalls for households

mt-digital opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3439)
<!-- Reviewable:end -->

mt-digital committed:
charging at start of iteration in Residential stalls for households

haitamlaarabi reviewed (changes requested):
> Reviewed 4 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3439)* status: 4 of 7 files reviewed, 5 unresolved discussions (waiting on @haitamlaarabi and @mt-digital)

___
*[src/main/scala/beam/agentsim/infrastructure/InfrastructureFunctions.scala, line 164 at r1](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3439#-MtBByMiAp_jbpD65554:-MtBByMiAp_jbpD65555:bl8wc7z) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/98ca46b2ac41c7d5f711dcea7818f103625aa530/src/main/scala/beam/agentsim/infrastructure/InfrastructureFunctions.scala#L164)):*
> ```Scala
>       inquiry,
>       ParkingZoneSearch.incrementalParkingZoneSearch(
>         // inquiry,
> ```

better if we remove this comment
___
*[src/main/scala/beam/agentsim/infrastructure/ParkingStall.scala, line 112 at r1](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3439#-MtBCJpMBQtKOihoDgl_:-MtBCJpMBQtKOihoDgla:b-8zpo21) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/98ca46b2ac41c7d5f711dcea7818f103625aa530/src/main/scala/beam/agentsim/infrastructure/ParkingStall.scala#L112)):*
> ```Scala
>     defaultGeoId: Id[_],
>     parkingZoneId: Id[ParkingZoneId] = Id.create("default", classOf[ParkingZoneId]),
>     chargingPointType: Option[ChargingPointType] = None
> ```

what's the role of these additions. I don't seen them being used.
___
*[src/main/scala/beam/agentsim/infrastructure/parking/ParkingZone.scala, line 126 at r1](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3439#-MtBCTtQ0B-Y8gPqeyVm:-MtBCTtQ0B-Y8gPqeyVn:b-qhuuoo) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/98ca46b2ac41c7d5f711dcea7818f103625aa530/src/main/scala/beam/agentsim/infrastructure/parking/ParkingZone.scala#L126)):*
> ```Scala
>     parkingType: ParkingType,
>     numStalls: Int,
>     chargingPointType: Option[ChargingPointType] = None
> ```

Same here. What's the role of this addition?
___
*[src/main/scala/beam/agentsim/infrastructure/parking/ParkingZoneSearch.scala, line 5 at r1](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3439#-MtBCacU4fv0hf-use1j:-MtBCacU4fv0hf-use1k:b-2guby3) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/98ca46b2ac41c7d5f711dcea7818f103625aa530/src/main/scala/beam/agentsim/infrastructure/parking/ParkingZoneSearch.scala#L5)):*
> ```Scala
> import beam.agentsim.agents.choice.logit.MultinomialLogit
> import beam.agentsim.infrastructure.ParkingStall
> import beam.agentsim.infrastructure.parking.ParkingZone._
> ```

same here
___
*[src/test/scala/beam/integration/ChargingSpec.scala, line 217 at r1](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3439#-MtBEmp1990iUiEbPmxc:-MtBEmp1990iUiEbPmxd:b-4nptc6) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/98ca46b2ac41c7d5f711dcea7818f103625aa530/src/test/scala/beam/integration/ChargingSpec.scala#L217)):*
> ```Scala
>     //
>     // Check that there is a charging event for start of simulation for both iterations.
>     val fuelSeshsAtTick0 = chargingPlugInEvents.filter(event => event._1 == 0)
> ```

As I said during the meeting. No need for nIter, instead at least one ChargingPlugIn event, and then the corresponding CharginRefuelSession has to contain some fuel


<!-- Sent from Reviewable.io -->

mt-digital committed:
removed unused code

mt-digital committed:
updated ChargingSpec according to HL comments on PR

mt-digital reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3439)* status: 1 of 7 files reviewed, 5 unresolved discussions (waiting on @haitamlaarabi)

___
*[src/test/scala/beam/integration/ChargingSpec.scala, line 217 at r1](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3439#-MtBEmp1990iUiEbPmxc:-MtKKYjI28dupxpMY54v:bwsk8an) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/98ca46b2ac41c7d5f711dcea7818f103625aa530/src/test/scala/beam/integration/ChargingSpec.scala#L217)):*
<details><summary><i>Previously, haitamlaarabi (Haitam Laarabi) wrote…</i></summary><blockquote>

As I said during the meeting. No need for nIter, instead at least one ChargingPlugIn event, and then the corresponding CharginRefuelSession has to contain some fuel
</blockquote></details>

OK, removed use of nIter; now checking for at least one PlugIn event. I added a line to the test config to initialize charge at 0.2*capacity on average as you recommended. It appears there is already a check at line 208 that plug/refuel events are consistent, and this should include the start of iteration charging. Of course please lmk if that's mistaken.
___
*[src/main/scala/beam/agentsim/infrastructure/InfrastructureFunctions.scala, line 164 at r1](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3439#-MtBByMiAp_jbpD65554:-MtKChHrALqn0KnOFGdA:b-whlh40) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/98ca46b2ac41c7d5f711dcea7818f103625aa530/src/main/scala/beam/agentsim/infrastructure/InfrastructureFunctions.scala#L164)):*
<details><summary><i>Previously, haitamlaarabi (Haitam Laarabi) wrote…</i></summary><blockquote>

better if we remove this comment
</blockquote></details>

Agreed, sorry about that.
___
*[src/main/scala/beam/agentsim/infrastructure/ParkingStall.scala, line 112 at r1](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3439#-MtBCJpMBQtKOihoDgl_:-MtKCqoD7cisVKfrMfU5:bifqsql) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/98ca46b2ac41c7d5f711dcea7818f103625aa530/src/main/scala/beam/agentsim/infrastructure/ParkingStall.scala#L112)):*
<details><summary><i>Previously, haitamlaarabi (Haitam Laarabi) wrote…</i></summary><blockquote>

what's the role of these additions. I don't seen them being used.
</blockquote></details>

You are correct. These have been removed in latest push to the branch.
___
*[src/main/scala/beam/agentsim/infrastructure/parking/ParkingZone.scala, line 126 at r1](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3439#-MtBCTtQ0B-Y8gPqeyVm:-MtKCx8cByc9Aaup5mYF:bpagrha) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/98ca46b2ac41c7d5f711dcea7818f103625aa530/src/main/scala/beam/agentsim/infrastructure/parking/ParkingZone.scala#L126)):*
<details><summary><i>Previously, haitamlaarabi (Haitam Laarabi) wrote…</i></summary><blockquote>

Same here. What's the role of this addition?
</blockquote></details>

It was unnecessary, apologies again.
___
*[src/main/scala/beam/agentsim/infrastructure/parking/ParkingZoneSearch.scala, line 5 at r1](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3439#-MtBCacU4fv0hf-use1j:-MtKD-9REGcZ89QX2AaL:bz2ydb0) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/98ca46b2ac41c7d5f711dcea7818f103625aa530/src/main/scala/beam/agentsim/infrastructure/parking/ParkingZoneSearch.scala#L5)):*
<details><summary><i>Previously, haitamlaarabi (Haitam Laarabi) wrote…</i></summary><blockquote>

same here
</blockquote></details>

Again, you are correct, this has been removed.


<!-- Sent from Reviewable.io -->

mt-digital commented:
___
*[src/test/scala/beam/integration/ChargingSpec.scala, line 217 at r1](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3439#-MtBEmp1990iUiEbPmxc:-MtKN_6E8wEOqznGtUQp:bceh8gn) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/98ca46b2ac41c7d5f711dcea7818f103625aa530/src/test/scala/beam/integration/ChargingSpec.scala#L217)):*
<details><summary><i>Previously, mt-digital (Matt Turner) wrote…</i></summary><blockquote>

OK, removed use of nIter and checking for at least one PlugIn event. I added a line to the test config to initialize charge at 0.2*capacity on average as you recommended. It appears there is already a check at line 208 that plug/refuel events are consistent, and this should include the start of iteration charging. Of course please lmk if that's mistaken.
</blockquote></details>

Sorry, correction, it's line 190 that tests plug/refuel events are consistent in ChargingSpec.


<!-- Sent from Reviewable.io -->

haitamlaarabi commented:
Test!

haitamlaarabi committed:
adding overnight charging at work and CAV support

haitamlaarabi commented:
Test Now!

JustinPihony commented on src/main/scala/beam/agentsim/agents/household/HouseholdActor.scala:
```diff
@@ -187,48 +188,47 @@ object HouseholdActor {
-            }
+            person.getSelectedPlan.getPlanElements.asScala.headOption
+              .flatMap {
+                case act: Activity if List("Home", "Work").contains(act.getType) =>
```
I would suggest creating the `List("Home", "Work")` as a class value, otherwise you are constantly creating and throwing it away - which will hurt GC.

haitamlaarabi replied:
Done

JustinPihony commented on src/main/scala/beam/agentsim/agents/household/HouseholdActor.scala:
```diff
@@ -187,48 +188,47 @@ object HouseholdActor {
     private var cavPassengerSchedules: Map[BeamVehicle, PassengerSchedule] = Map()
     private var personAndActivityToCav: Map[(Id[Person], Activity), BeamVehicle] = Map()
     private var personAndActivityToLegs: Map[(Id[Person], Activity), List[BeamLeg]] = Map()
+    private var householdMembersAndTheirLocation: Map[Id[Person], (String, Coord)] = Map()
```
This should be renamed to `householdMembersToLocationTypeAndLocation` - I'm all about full transparency in variable naming. Maybe it could be FullyQualifiedLocation?

haitamlaarabi replied:
Done

haitamlaarabi committed:
addressing @JustinPihony

haitamlaarabi reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3439)* status: 0 of 7 files reviewed, 8 unresolved discussions (waiting on @haitamlaarabi, @JustinPihony, @mt-digital, and @zneedell)

___
*[src/main/scala/beam/agentsim/agents/household/HouseholdFleetManager.scala, line 138 at r4](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3439#-MvSPSeB84r9AvbRZ7LK:-MvSPSeB84r9AvbRZ7LL:bd5rp1t) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/0c64e35dd0c8889cb86f6e1416c91d432fde4248/src/main/scala/beam/agentsim/agents/household/HouseholdFleetManager.scala#L138)):*
> ```Scala
> 
>     case MobilityStatusInquiry(personId, whenWhere, _, requireVehicleCategoryAvailable, triggerId) =>
>       {
> ```

@zneedell It seems that @justinpihony  is not happy with this section of the code that has already been merged but we have chance to fix it in this PR. I tried to do minimal changes to not mess up the logic of your code. Here are his comment from a previous PR:

"You can change the if to be a match and then the Nil case moves from an unnecessary logging to being the primary emergency generation. Note that you could still end up with a None vehicle in the scenario where no available vehicles and no emergency generator is configured to be available. How you want to handle that (logging?) is up to you."
and 
"I am just now noticing that this is an Option[Unit], which means it is basically working off of the fact whether the for did something at all. Is that what is wanted? If so, then maybe it is better to assign to a variable and then do an if isEmpty. Ultimately it is more about confusing a reader of the code, so it will be functionally the same and not worth blocking., though"


<!-- Sent from Reviewable.io -->

haitamlaarabi committed:
adding a TODO

haitamlaarabi reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3439)* status: 0 of 8 files reviewed, 9 unresolved discussions (waiting on @haitamlaarabi, @JustinPihony, @mt-digital, and @zneedell)

___
*[src/test/scala/beam/agentsim/agents/PersonWithPersonalVehiclePlanSpec.scala, line 681 at r5](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3439#-MvSQijkBlmfXTq-5KRR:-MvSQijkBlmfXTq-5KRS:b-9g3t5x) ([raw file](https://github.com/lbnl-ucb-sti/beam/blob/3a973c42a45a64ab3d87f9676e79311b33c1ff66/src/test/scala/beam/agentsim/agents/PersonWithPersonalVehiclePlanSpec.scala#L681)):*
> ```Scala
>       expectMsgType[CompletionNotice]
> 
>       // TODO: Testing last resort vehicle creation
> ```

@zneedell here to. @justinpihony said:
"Per the name, this test should also be testing against the vehicle, shouldn't it?"
and
"I guess I was wondering if there is a way to create an assertion that a last resort car was created? A stub that could be verified that the emergency creation was triggered - or a test of the vehicle attributes?"


<!-- Sent from Reviewable.io -->

haitamlaarabi committed:
less HOV trace

haitamlaarabi committed:
moving away from Init to actual activities

haitamlaarabi committed:
adding fixes to overnight charging step1

haitamlaarabi committed:
merging with develop

haitamlaarabi committed:
adding fixed to overnight charging

haitamlaarabi committed:
renaming slow charging only function

haitamlaarabi committed:
adding if refuel needed to parking

haitamlaarabi committed:
filtering out Destination and Enroute Charging

haitamlaarabi committed:
adding more condition in park

haitamlaarabi committed:
merging with hl/gemini-develop-enroute-overnight-charging

haitamlaarabi committed:
merging with hl/gemini-develop-enroute-overnight-charging

haitamlaarabi committed:
fmt

haitamlaarabi committed:
fixing implicit call

haitamlaarabi commented:
Test now!

haitamlaarabi committed:
Testing last resort vehicle creation

haitamlaarabi commented:
Test now!

JustinPihony commented on src/main/scala/beam/utils/scenario/urbansim/HOVModeTransformer.scala:
```diff
@@ -225,9 +225,9 @@ object HOVModeTransformer extends LazyLogging {
     if (cantSplitTripsForPersons.nonEmpty) {
       logger.info(
-        "Cannot split plans to trips because plans does not start and end by Home activity for {} persons: {}",
+        "Cannot split plans to trips because plans does not start and end by Home activity for {} persons including: {}",
```
Should we say explicitly that it is "...including first 10"

JustinPihony commented on src/main/scala/beam/agentsim/infrastructure/ScaleUpCharging.scala:
```diff
@@ -53,6 +53,23 @@ trait ScaleUpCharging extends {
     }
   }
 
+//  lazy val nonSampledVehicle: Map[Id[BeamVehicle], BeamVehicle] = {
```
Does this have a purpose or can we delete it?

JustinPihony commented on src/main/scala/beam/agentsim/infrastructure/ChargingFunctions.scala:
```diff
@@ -136,11 +155,50 @@ class ChargingFunctions[GEO: GeoLevel](
+        case Some(beamVehicle) =>
+          parkingAlternative.parkingZone.chargingPointType match {
+            case Some(chargingPoint) =>
+              val (_, addedEnergy) = ChargingPointType.calculateChargingSessionLengthAndEnergyInJoule(
```
This is bound to create confusion since the outside variable name is `addedEnergy` also. I would suggest changing to something else on one end or the other.

JustinPihony commented on src/main/R/gemini/gemini-processing.R:
```diff
@@ -21,8 +21,82 @@ shpFile <- pp(workDir, "/shapefile/Oakland+Alameda+TAZ/Transportation_Analysis_Z
 oaklandCbg <- st_read(shpFile)
 
 ###
-#eventsraw <- readCsv(pp(workDir, "/2021Aug22-Oakland/BASE0/events-raw/2.events.BASE0.csv.gz"))
```
I did not dig into the R code as it is definitely a weak point of mine. If you feel comfortable then go ahead using my review without this code.

JustinPihony commented on src/main/python/nhts_analysis/get-NHTS-mode_choice.py:
```diff
@@ -0,0 +1,279 @@
+import pandas as pd
```
I also did not dig too far into the python, but gave a fair skim

JustinPihony commented on src/main/scala/beam/agentsim/agents/household/HouseholdActor.scala:
```diff
@@ -199,46 +206,55 @@ object HouseholdActor {
+            whoDrivesThisVehicle = whoDrivesThisVehicle + (vehicleIdFromPlans -> person.getId)
+          }
+          person.getSelectedPlan.getPlanElements.asScala.headOption
+            .flatMap {
```
This could be a `collect` which returns `person.getId -> (ParkingInquiry.activityTypeStringToEnum(act.getType), act.getType, act.getCoord)` directly

JustinPihony commented on src/main/scala/beam/agentsim/agents/household/HouseholdActor.scala:
```diff
@@ -189,6 +194,8 @@ object HouseholdActor {
     private var cavPassengerSchedules: Map[BeamVehicle, PassengerSchedule] = Map()
     private var personAndActivityToCav: Map[(Id[Person], Activity), BeamVehicle] = Map()
     private var personAndActivityToLegs: Map[(Id[Person], Activity), List[BeamLeg]] = Map()
+    private var householdMembersToLocationTypeAndLocation: Map[Id[Person], (ParkingActivityType, String, Coord)] = Map()
```
I realize there are some cases where this has been done already, but we really should be using a `mutable.Map` instead of `var`. https://beam.readthedocs.io/en/latest/developers.html#use-mutable-buffer-instead-of-immutable-var

JustinPihony reviewed (approved):
A few comments that should be addressed, but not necessarily worth blocking. Although the variable naming one is close.

JustinPihony committed:
Merge branch 'develop' into mt/start-of-iter-charging

haitamlaarabi commented:
This work has been merged under another PR. I'm closing this PR!
